File: src/speedTest.js
const { brotliDecompress, zlibDecompress, loadFile } = require("./utils");

const main = async () => {
  const zlibCompression = await loadFile("../../data/zlibBatchData.json")
  const brotliCompression = await loadFile("../../data/brotliBatchData.json")
  const maxBrotliCompression = await loadFile("../../data/maxBrotliCompressionData.json")
  const minBrotliCompression = await loadFile("../../data/minBrotliCompressionData.json")
  const maxZlibCompression = await loadFile("../../data/maxZlibCompressionData.json")
  const minZlibCompression = await loadFile("../../data/minZlibCompressionData.json")

  // Loop Count
  const loopCount = 5;

  // Time
  let totalTime = 0;
  let avgTime = 0;

  for (let i = 0; i < loopCount; i++) {
    start = new Date().getTime();

    for (const payload of brotliCompression) {
      brotliDecompress(payload);
    }

    end = new Date().getTime();
    totalTime += end - start;

    console.log(`Brotli Decompression Execution Time: ${end - start} ms | Loop: ${i + 1} time`);
  }
  console.log(`Brotli Decompression Speed Test: ${totalTime / loopCount} ms | Loop: ${loopCount}`);

  totalTime = 0;
  avgTime = 0;

  for (let i = 0; i < loopCount; i++) {
    start = new Date().getTime();

    for (const payload of zlibCompression) {
      zlibDecompress(payload);
    }

    end = new Date().getTime();
    totalTime += end - start;

    console.log(`Brotli Decompression Execution Time: ${end - start} ms | Loop: ${i + 1} time`);
  }
  console.log(`Brotli Decompression Speed Test: ${totalTime / loopCount} ms | Loop: ${loopCount}`);

  totalTime = 0;
  avgTime = 0;

  for (let i = 0; i < loopCount; i++) {
    start = new Date().getTime();

    for (const payload of maxBrotliCompression) {
      brotliDecompress(payload);
    }

    end = new Date().getTime();
    totalTime += end - start;

    console.log(`Brotli Decompression Execution Time: ${end - start} ms | Loop: ${i + 1} time (worst case)`);
  }
  console.log(`Brotli Decompression Speed Test: ${totalTime / loopCount} ms | Loop: ${loopCount} (worst case)`);

  totalTime = 0;
  avgTime = 0;

  for (let i = 0; i < loopCount; i++) {
    start = new Date().getTime();

    for (const payload of minBrotliCompression) {
      brotliDecompress(payload);
    }

    end = new Date().getTime();
    totalTime += end - start;

    console.log(`Brotli Decompression Execution Time: ${end - start} ms | Loop: ${i + 1} time (best case)`);
  }
  console.log(`Brotli Decompression Speed Test: ${totalTime / loopCount} ms | Loop: ${loopCount} (best case)`);

  totalTime = 0;
  avgTime = 0;

  for (let i = 0; i < loopCount; i++) {
    start = new Date().getTime();

    for (const payload of maxZlibCompression) {
      zlibDecompress(payload);
    }

    end = new Date().getTime();
    totalTime += end - start;

    console.log(`Brotli Decompression Execution Time: ${end - start} ms | Loop: ${i + 1} time (worst case)`);
  }
  console.log(`Brotli Decompression Speed Test: ${totalTime / loopCount} ms | Loop: ${loopCount} (worst case)`);

  totalTime = 0;
  avgTime = 0;

  for (let i = 0; i < loopCount; i++) {
    start = new Date().getTime();

    for (const payload of minZlibCompression) {
        zlibDecompress(payload);
    }

    end = new Date().getTime();
    totalTime += end - start;

    console.log(`Brotli Decompression Execution Time: ${end - start} ms | Loop: ${i + 1} time (best case)`);
  }
  console.log(`Brotli Decompression Speed Test: ${totalTime / loopCount} ms | Loop: ${loopCount} (best case)`);

};

main();

File: go/speedTest.go
package main

import (
	"bytes"
	"compress/zlib"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"time"

	"github.com/shirou/gopsutil/v3/cpu"
	"github.com/shirou/gopsutil/v3/host"
	"github.com/shirou/gopsutil/v3/mem"

	"github.com/andybalholm/brotli"
)

func main() {
	type TransactionsData []string

	type SysInfo struct {
		Hostname string
		Platform string
		CPU      string
		Core     int32
		RAM      uint64
	}

	hostStat, _ := host.Info()
	cpuStat, _ := cpu.Info()
	vmStat, _ := mem.VirtualMemory()

	info := new(SysInfo)

	info.Hostname = hostStat.Hostname
	info.Platform = hostStat.Platform
	info.CPU = cpuStat[0].ModelName
	info.Core = cpuStat[0].Cores
	info.RAM = vmStat.Total / 1024 / 1024

	// System Info
	infoJSON, err := json.MarshalIndent(info, "", "  ")
	if err != nil {
		panic("Failed to marshal: " + err.Error())
	}
	fmt.Printf("%s\n", string(infoJSON))

	// Read file
	data, err := ioutil.ReadFile("../data/transactionsData.json")
	if err != nil {
		panic("Failed to load data: " + err.Error())
	}

	var transactionsData TransactionsData
	err = json.Unmarshal(data, &transactionsData)
	if err != nil {
		panic("Failed to decode json: " + err.Error())
	}

	// Loop Count
	loopCount := 5

	// Time Variables
	var totalTime time.Duration
	var avgTime time.Duration

	fmt.Println("\n-------------------------------------")

	fmt.Println("\nTotal TXs:", len(transactionsData), " | Loop: ", loopCount)

	fmt.Println("\n-------------------------------------")

	fmt.Println("\nZlib Speed Test:")

	for i := 0; i < loopCount; i++ {
		start := time.Now()

		for _, transaction := range transactionsData {
			var in bytes.Buffer
			b := []byte(transaction)
			w := zlib.NewWriter(&in)
			w.Write(b)
			w.Close()
		}

		end := time.Now()
		totalTime += end.Sub(start)
	}
	fmt.Println("\nTotal Time: ", totalTime.String(), " | Loop: ", loopCount)
	avgTime = totalTime / time.Duration(loopCount)
	fmt.Println("Average Time:", avgTime)
	fmt.Println("\n-------------------------------------")

	totalTime = time.Duration(0)
	avgTime = time.Duration(0)

	fmt.Println("\nBrotli Speed Test:")

	for i := 0; i < loopCount; i++ {
		start := time.Now()

		for _, transaction := range transactionsData {
			var in bytes.Buffer
			b := []byte(transaction)
			w := brotli.NewWriterLevel(&in, 11)
			w.Write(b)
			w.Close()
		}

		end := time.Now()
		totalTime += end.Sub(start)

		fmt.Println("Execution Duration: ", end.Sub(start), " | Loop: ", i+1, " time")
	}
	fmt.Println("\nTotal Time: ", totalTime.String(), " | Loop: ", loopCount)
	avgTime = totalTime / time.Duration(loopCount)
	fmt.Println("Average Time:", avgTime)
	fmt.Println("\n-------------------------------------")
}
